include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -223,108 +218,155 @@ class NBL_API2 IShaderCompiler : public core::IReferenceCounted
+						
+						otherSortedExtraDefines.assign(other.extraDefines.begin(), other.extraDefines.end());
+						std::sort(otherSortedExtraDefines.begin(), otherSortedExtraDefines.end(), [](const SMacroData& lhs, const SMacroData& rhs) {return lhs.identifier < rhs.identifier; });
+
+						for (auto definesIt = sortedExtraDefines.begin(), otherDefinesIt = otherSortedExtraDefines.begin(); definesIt != sortedExtraDefines.end(); definesIt++, otherDefinesIt++) {
 							if (definesIt->identifier != otherDefinesIt->identifier || definesIt->definition != otherDefinesIt->definition) return false;
 						}
 						return true;
 					}
 
-					// We don't need to hash the preprocessor on its own but in conjunction with the compiler options. We return data to perform the hash on from SCompilerOptions
-					inline std::vector<char> getHashable() const
+				private:
+
+					// Only SCompilerArgs should instantiate this struct
+					SPreprocessorArgs(const SPreprocessorOptions& options)
+						: sourceIdentifier(options.sourceIdentifier)
 					{
-						std::vector<char> hashable;
-						hashable.insert(hashable.end(), sourceIdentifier.data()[0], sourceIdentifier.data()[sourceIdentifier.size()]);
-						core::vector<SMacroData> sortedExtraDefines;
-						sortedExtraDefines.assign(extraDefines.begin(), extraDefines.end());
-						// Sort them by identifier so the hash is not order-sensitive!
-						std::sort(sortedExtraDefines.begin(), sortedExtraDefines.end(), [](const SMacroData& lhs, const SMacroData& rhs) {return lhs.identifier < rhs.identifier; });
-						for (const auto& defines : sortedExtraDefines) {
-							hashable.insert(hashable.end(), defines.identifier.data()[0], defines.identifier.data()[defines.identifier.size()]);
-							hashable.insert(hashable.end(), defines.definition.data()[0], defines.definition.data()[defines.definition.size()]);
+						for (auto define : options.extraDefines) {
+							SMacroData data{ std::string(define.identifier), std::string(define.definition) };
+							extraDefines.push_back(data);
 						}
-
-						return hashable;
-					}
-
+					};
 					std::string sourceIdentifier;
 					std::vector<SMacroData> extraDefines;
+					NLOHMANN_DEFINE_TYPE_INTRUSIVE(SPreprocessorArgs, sourceIdentifier, extraDefines);
 				};
 
-				struct SCompilerData {
+				struct SCompilerArgs final {
 
-					inline bool operator==(const SCompilerData& other) const {
-						if (stage != other.stage || targetSpirvVersion != other.targetSpirvVersion || debugInfoFlags != other.debugInfoFlags || preprocessorData != other.preprocessorData) return false;
-						if (optimizerPasses.size() != other.optimizerPasses.size()) return false;
+					friend class SEntry;
+					friend void to_json(nlohmann::json&, const SCompilerArgs&);
+					friend void from_json(const nlohmann::json&, SCompilerArgs&);
+					// Public default needed for json serialization
+					SCompilerArgs() {}
```

Comment:
which method exactly?

Reply:
The unordered multiset gets stored by the json library as an array (all iterables get stored as a json array). To be able to retrieve a struct from an array the library seems to require a default constructor. `SEntry` is actually the one that needs to have a (public) default constructor, but that requires SCompilerArgs to have one as well. I could alternatively make only `SEntry` have default public constructor and make this one be private, since `SEntry` is friend